File: backend/src/main/java/com/pgvaale/backend/service/TiffinService.java
package com.pgvaale.backend.service;

import com.pgvaale.backend.dto.MenuDTO;
import com.pgvaale.backend.dto.TiffinDashboardDTO;
import com.pgvaale.backend.dto.UserTiffinDTO;
import com.pgvaale.backend.entity.Menu;
import com.pgvaale.backend.entity.Tiffin;
import com.pgvaale.backend.entity.User;
import com.pgvaale.backend.entity.UserTiffin;
import com.pgvaale.backend.repository.MenuRepository;
import com.pgvaale.backend.repository.TiffinRepository;
import com.pgvaale.backend.repository.UserRepository;
import com.pgvaale.backend.repository.UserTiffinRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class TiffinService {
    @Autowired
    private TiffinRepository tiffinRepository;
    
    @Autowired
    private MenuRepository menuRepository;
    
    @Autowired
    private UserTiffinRepository userTiffinRepository;
    
    @Autowired
    private UserRepository userRepository;

    public List<Tiffin> getAllTiffins() {
        return tiffinRepository.findAll();
    }

    public Optional<Tiffin> getTiffinById(Long id) {
        return tiffinRepository.findById(id);
    }

    public Tiffin saveTiffin(Tiffin tiffin) {
        return tiffinRepository.save(tiffin);
    }

    public void deleteTiffin(Long id) {
        tiffinRepository.deleteById(id);
    }
    
    // Menu Management
    public MenuDTO createMenu(MenuDTO menuDTO) {
        Tiffin tiffin = tiffinRepository.findById(menuDTO.getTiffinId())
                .orElseThrow(() -> new RuntimeException("Tiffin not found"));
        
        Menu menu = Menu.builder()
                .tiffin(tiffin)
                .dayOfWeek(menuDTO.getDayOfWeek())
                .breakfast(menuDTO.getBreakfast())
                .lunch(menuDTO.getLunch())
                .dinner(menuDTO.getDinner())
                .menuDate(menuDTO.getMenuDate())
                .foodCategory(menuDTO.getFoodCategory())
                .price(menuDTO.getPrice())
                .isActive(true)
                .build();
        
        Menu savedMenu = menuRepository.save(menu);
        return convertToDTO(savedMenu);
    }
    
    public MenuDTO updateMenu(Long menuId, MenuDTO menuDTO) {
        Menu menu = menuRepository.findById(menuId)
                .orElseThrow(() -> new RuntimeException("Menu not found"));
        
        menu.setBreakfast(menuDTO.getBreakfast());
        menu.setLunch(menuDTO.getLunch());
        menu.setDinner(menuDTO.getDinner());
        menu.setMenuDate(menuDTO.getMenuDate());
        menu.setFoodCategory(menuDTO.getFoodCategory());
        menu.setPrice(menuDTO.getPrice());
        
        Menu updatedMenu = menuRepository.save(menu);
        return convertToDTO(updatedMenu);
    }
    
    public void deleteMenu(Long menuId) {
        Menu menu = menuRepository.findById(menuId)
                .orElseThrow(() -> new RuntimeException("Menu not found"));
        menu.setActive(false);
        menuRepository.save(menu);
    }
    
    public List<MenuDTO> getWeeklyMenu(Long tiffinId) {
        List<Menu> menus = menuRepository.findByTiffinIdAndIsActiveTrue(tiffinId);
        return menus.stream().map(this::convertToDTO).collect(Collectors.toList());
    }
    
    public MenuDTO getMenuByDay(Long tiffinId, String dayOfWeek) {
        Optional<Menu> menu = menuRepository.findByTiffinIdAndDayOfWeekAndIsActiveTrue(tiffinId, dayOfWeek);
        return menu.map(this::convertToDTO).orElse(null);
    }
    
    // User Request Management
    public UserTiffinDTO createUserRequest(Long userId, Long tiffinId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        Tiffin tiffin = tiffinRepository.findById(tiffinId)
                .orElseThrow(() -> new RuntimeException("Tiffin not found"));
        
        // Check if request already exists
        Optional<UserTiffin> existingRequest = userTiffinRepository.findByUserIdAndTiffinId(userId, tiffinId);
        if (existingRequest.isPresent()) {
            throw new RuntimeException("Request already exists");
        }
        
        UserTiffin userTiffin = UserTiffin.builder()
                .user(user)
                .tiffin(tiffin)
                .status(UserTiffin.RequestStatus.PENDING)
                .assignedDateTime(LocalDateTime.now())
                .build();
        
        UserTiffin savedUserTiffin = userTiffinRepository.save(userTiffin);
        return UserTiffinDTO.fromEntity(savedUserTiffin);
    }
    
    public UserTiffinDTO updateRequestStatus(Long requestId, UserTiffin.RequestStatus status) {
        UserTiffin userTiffin = userTiffinRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));
        
        userTiffin.setStatus(status);
        if (status == UserTiffin.RequestStatus.REJECTED) {
            userTiffin.setDeletionDateTime(LocalDateTime.now());
        }
        
        UserTiffin updatedUserTiffin = userTiffinRepository.save(userTiffin);
        return UserTiffinDTO.fromEntity(updatedUserTiffin);
    }
    
    public List<UserTiffinDTO> getTiffinRequests(Long tiffinId, UserTiffin.RequestStatus status) {
        List<UserTiffin> userTiffins;
        if (status == null) {
            userTiffins = userTiffinRepository.findByTiffinId(tiffinId);
        } else {
            userTiffins = userTiffinRepository.findByTiffinIdAndStatusOrderByAssignedDateTimeDesc(tiffinId, status);
        }
        return userTiffins.stream().map(UserTiffinDTO::fromEntity).collect(Collectors.toList());
    }
    
    public List<UserTiffinDTO> getUserRequests(Long userId, UserTiffin.RequestStatus status) {
        List<UserTiffin> userTiffins;
        if (status == null) {
            userTiffins = userTiffinRepository.findByUserId(userId);
        } else {
            userTiffins = userTiffinRepository.findByUserIdAndStatusOrderByAssignedDateTimeDesc(userId, status);
        }
        return userTiffins.stream().map(UserTiffinDTO::fromEntity).collect(Collectors.toList());
    }
    
    public void cancelUserRequest(Long requestId) {
        UserTiffin userTiffin = userTiffinRepository.findById(requestId)
                .orElseThrow(() -> new RuntimeException("Request not found"));
        
        userTiffin.setStatus(UserTiffin.RequestStatus.REJECTED);
        userTiffin.setDeletionDateTime(LocalDateTime.now());
        userTiffinRepository.save(userTiffin);
    }
    
    // Dashboard
    public TiffinDashboardDTO getTiffinDashboard(Long tiffinId) {
        Tiffin tiffin = tiffinRepository.findById(tiffinId)
                .orElseThrow(() -> new RuntimeException("Tiffin not found"));
        
        Long pendingRequests = userTiffinRepository.countByTiffinIdAndStatus(tiffinId, UserTiffin.RequestStatus.PENDING);
        Long acceptedRequests = userTiffinRepository.countByTiffinIdAndStatus(tiffinId, UserTiffin.RequestStatus.ACCEPTED);
        Long rejectedRequests = userTiffinRepository.countByTiffinIdAndStatus(tiffinId, UserTiffin.RequestStatus.REJECTED);
        
        List<UserTiffinDTO> recentRequests = userTiffinRepository.findByTiffinId(tiffinId)
                .stream()
                .limit(5)
                .map(UserTiffinDTO::fromEntity)
                .collect(Collectors.toList());
        
        return TiffinDashboardDTO.builder()
                .tiffinName(tiffin.getName())
                .pendingRequests(pendingRequests)
                .acceptedRequests(acceptedRequests)
                .rejectedRequests(rejectedRequests)
                .averageRating(0.0) // TODO: Implement rating system
                .recentRequests(recentRequests)
                .build();
    }
    
    private MenuDTO convertToDTO(Menu menu) {
        return MenuDTO.builder()
                .id(menu.getId())
                .tiffinId(menu.getTiffin().getId())
                .dayOfWeek(menu.getDayOfWeek())
                .breakfast(menu.getBreakfast())
                .lunch(menu.getLunch())
                .dinner(menu.getDinner())
                .menuDate(menu.getMenuDate())
                .foodCategory(menu.getFoodCategory())
                .price(menu.getPrice())
                .isActive(menu.isActive())
                .build();
    }
} 

File: PGVaaleDotNetBackend/Services/TiffinService.cs
using PGVaaleDotNetBackend.DTOs;
using PGVaaleDotNetBackend.Entities;
using PGVaaleDotNetBackend.Repositories;

namespace PGVaaleDotNetBackend.Services
{
    public class TiffinService
    {
        private readonly ITiffinRepository _tiffinRepository;
        private readonly IMenuRepository _menuRepository;
        private readonly IUserTiffinRepository _userTiffinRepository;
        private readonly IUserRepository _userRepository;
        private readonly IFeedback_TiffinRepository _feedbackTiffinRepository;

        public TiffinService(
            ITiffinRepository tiffinRepository,
            IMenuRepository menuRepository,
            IUserTiffinRepository userTiffinRepository,
            IUserRepository userRepository,
            IFeedback_TiffinRepository feedbackTiffinRepository)
        {
            _tiffinRepository = tiffinRepository;
            _menuRepository = menuRepository;
            _userTiffinRepository = userTiffinRepository;
            _userRepository = userRepository;
            _feedbackTiffinRepository = feedbackTiffinRepository;
        }

        // CRUD Operations
        public IEnumerable<Tiffin> GetAllTiffins()
        {
            return _tiffinRepository.GetAll();
        }

        public async Task<IEnumerable<Tiffin>> GetAllTiffinsAsync()
        {
            return await Task.FromResult(_tiffinRepository.GetAll());
        }

        public Tiffin? GetTiffinById(long id)
        {
            return _tiffinRepository.GetById(id);
        }

        public Tiffin SaveTiffin(Tiffin tiffin)
        {
            if (tiffin.Id == 0)
            {
                _tiffinRepository.Add(tiffin);
            }
            else
            {
                _tiffinRepository.Update(tiffin);
            }
            return tiffin;
        }

        public void DeleteTiffin(long id)
        {
            _tiffinRepository.Delete(id);
        }

        // Menu Management
        public MenuDTO CreateMenu(MenuDTO menuDTO)
        {
            var tiffin = _tiffinRepository.GetById(menuDTO.TiffinId);
            if (tiffin == null)
                throw new InvalidOperationException("Tiffin not found");

            var menu = new Menu
            {
                TiffinId = menuDTO.TiffinId,
                DayOfWeek = menuDTO.DayOfWeek,
                Breakfast = menuDTO.Breakfast,
                Lunch = menuDTO.Lunch,
                Dinner = menuDTO.Dinner,
                MenuDate = menuDTO.MenuDate,
                FoodCategory = menuDTO.FoodCategory,
                Price = menuDTO.Price,
                IsActive = true
            };

            _menuRepository.Add(menu);
            return ConvertToDTO(menu);
        }

        public MenuDTO UpdateMenu(long menuId, MenuDTO menuDTO)
        {
            var menu = _menuRepository.GetById(menuId);
            if (menu == null)
                throw new InvalidOperationException("Menu not found");

            menu.Breakfast = menuDTO.Breakfast;
            menu.Lunch = menuDTO.Lunch;
            menu.Dinner = menuDTO.Dinner;
            menu.MenuDate = menuDTO.MenuDate;
            menu.FoodCategory = menuDTO.FoodCategory;
            menu.Price = menuDTO.Price;

            _menuRepository.Update(menu);
            return ConvertToDTO(menu);
        }

        public void DeleteMenu(long menuId)
        {
            var menu = _menuRepository.GetById(menuId);
            if (menu == null)
                throw new InvalidOperationException("Menu not found");

            menu.IsActive = false;
            _menuRepository.Update(menu);
        }

        public async Task<IEnumerable<MenuDTO>> GetWeeklyMenuAsync(long tiffinId)
        {
            var menus = await _menuRepository.GetByTiffinIdAndIsActiveAsync(tiffinId, true);
            return menus.Select(ConvertToDTO);
        }

        public async Task<MenuDTO?> GetMenuByDayAsync(long tiffinId, string dayOfWeek)
        {
            var menu = await _menuRepository.GetByTiffinIdAndDayOfWeekAndIsActiveAsync(tiffinId, dayOfWeek, true);
            return menu != null ? ConvertToDTO(menu) : null;
        }

        // User Request Management
        public UserTiffinDTO CreateUserRequest(long userId, long tiffinId)
        {
            var user = _userRepository.GetById(userId);
            if (user == null)
                throw new InvalidOperationException("User not found");

            var tiffin = _tiffinRepository.GetById(tiffinId);
            if (tiffin == null)
                throw new InvalidOperationException("Tiffin not found");

            // Check if request already exists
            var existingRequest = _userTiffinRepository.GetByUserIdAndTiffinIdAsync(userId, tiffinId).Result;
            if (existingRequest != null)
                throw new InvalidOperationException("Request already exists");

            var userTiffin = new UserTiffin
            {
                UserId = userId,
                TiffinId = tiffinId,
                Status = UserTiffin.RequestStatus.PENDING,
                AssignedDateTime = DateTime.UtcNow
            };

            _userTiffinRepository.Add(userTiffin);
            return UserTiffinDTO.FromEntity(userTiffin);
        }

        public UserTiffinDTO UpdateRequestStatus(long requestId, UserTiffin.RequestStatus status)
        {
            var userTiffin = _userTiffinRepository.GetById(requestId);
            if (userTiffin == null)
                throw new InvalidOperationException("Request not found");

            userTiffin.Status = status;
            if (status == UserTiffin.RequestStatus.REJECTED)
            {
                userTiffin.DeletionDateTime = DateTime.UtcNow;
            }

            _userTiffinRepository.Update(userTiffin);
            return UserTiffinDTO.FromEntity(userTiffin);
        }

        public async Task<IEnumerable<UserTiffinDTO>> GetTiffinRequestsAsync(long tiffinId, UserTiffin.RequestStatus? status)
        {
            IEnumerable<UserTiffin> userTiffins;
            if (status == null)
            {
                userTiffins = await _userTiffinRepository.GetByTiffinIdAsync(tiffinId);
            }
            else
            {
                userTiffins = await _userTiffinRepository.GetByTiffinIdAndStatusAsync(tiffinId, status.Value);
            }
            return userTiffins.Select(UserTiffinDTO.FromEntity);
        }

        public async Task<IEnumerable<UserTiffinDTO>> GetUserRequestsAsync(long userId, UserTiffin.RequestStatus? status)
        {
            IEnumerable<UserTiffin> userTiffins;
            if (status == null)
            {
                userTiffins = await _userTiffinRepository.GetByUserIdAsync(userId);
            }
            else
            {
                userTiffins = await _userTiffinRepository.GetByUserIdAndStatusAsync(userId, status.Value);
            }
            return userTiffins.Select(UserTiffinDTO.FromEntity);
        }

        public void CancelUserRequest(long requestId)
        {
            var userTiffin = _userTiffinRepository.GetById(requestId);
            if (userTiffin == null)
                throw new InvalidOperationException("Request not found");

            userTiffin.Status = UserTiffin.RequestStatus.REJECTED;
            userTiffin.DeletionDateTime = DateTime.UtcNow;
            _userTiffinRepository.Update(userTiffin);
        }

        // Dashboard
        public async Task<TiffinDashboardDTO> GetTiffinDashboardAsync(long tiffinId)
        {
            var tiffin = _tiffinRepository.GetById(tiffinId);
            if (tiffin == null)
                throw new InvalidOperationException("Tiffin not found");

            var pendingRequests = await _userTiffinRepository.CountByTiffinIdAndStatusAsync(tiffinId, UserTiffin.RequestStatus.PENDING);
            var acceptedRequests = await _userTiffinRepository.CountByTiffinIdAndStatusAsync(tiffinId, UserTiffin.RequestStatus.ACCEPTED);
            var rejectedRequests = await _userTiffinRepository.CountByTiffinIdAndStatusAsync(tiffinId, UserTiffin.RequestStatus.REJECTED);

            var recentRequests = await _userTiffinRepository.GetByTiffinIdAsync(tiffinId);
            var recentRequestsDTO = recentRequests.Take(5).Select(UserTiffinDTO.FromEntity).ToList();

            // Calculate average rating for this tiffin
            var feedbacks = await _feedbackTiffinRepository.GetByTiffinIdAsync(tiffinId);
            var averageRating = feedbacks.Any() ? feedbacks.Average(f => f.Rating) : 0.0;

            return new TiffinDashboardDTO
            {
                TiffinName = tiffin.Name,
                PendingRequests = pendingRequests,
                AcceptedRequests = acceptedRequests,
                RejectedRequests = rejectedRequests,
                AverageRating = Math.Round(averageRating, 1),
                RecentRequests = recentRequestsDTO
            };
        }

        // Tiffin Feedback
        public Feedback_Tiffin SubmitTiffinFeedback(long userId, long tiffinId, int rating, string feedback)
        {
            var user = _userRepository.GetById(userId);
            if (user == null)
                throw new InvalidOperationException("User not found");

            var tiffin = _tiffinRepository.GetById(tiffinId);
            if (tiffin == null)
                throw new InvalidOperationException("Tiffin not found");

            var feedbackTiffin = new Feedback_Tiffin
            {
                UserId = userId,
                TiffinId = tiffinId,
                Rating = rating,
                Feedback = feedback,
                CreatedAt = DateTime.UtcNow
            };

            _feedbackTiffinRepository.Add(feedbackTiffin);
            return feedbackTiffin;
        }

        public async Task<IEnumerable<Feedback_Tiffin>> GetTiffinFeedbackByUserAsync(long userId)
        {
            return await _feedbackTiffinRepository.GetByUserIdAsync(userId);
        }

        public async Task<IEnumerable<Feedback_Tiffin>> GetTiffinFeedbackByTiffinAsync(long tiffinId)
        {
            return await _feedbackTiffinRepository.GetByTiffinIdAsync(tiffinId);
        }

        private MenuDTO ConvertToDTO(Menu menu)
        {
            return new MenuDTO
            {
                Id = menu.Id,
                TiffinId = menu.TiffinId,
                DayOfWeek = menu.DayOfWeek,
                Breakfast = menu.Breakfast,
                Lunch = menu.Lunch,
                Dinner = menu.Dinner,
                MenuDate = menu.MenuDate,
                FoodCategory = menu.FoodCategory,
                Price = menu.Price,
                IsActive = menu.IsActive
            };
        }
    }
}